crates/sp-executor/src/lib.rs

```diff
@@ -99,6 +115,213 @@ impl<Hash: Encode> BundleHeader<Hash> {
     }
 }
 
+// TODO: unify the type between the solution and threshold better.
+pub fn derive_bundle_election_solution(domain_id: DomainId, vrf_output: &[u8]) -> u64 {
+    let local_domain_randomness = blake2b_256_hash_list(&[&domain_id.to_le_bytes(), vrf_output]);
+
+    let election_solution = u64::from_le_bytes(
+        local_domain_randomness
+            .split_at(core::mem::size_of::<u64>())
+            .0
+            .try_into()
+            .expect("Local domain randomness must fit into u64; qed"),
+    );
+
+    election_solution
+}
+
+/// Returns the election threshold based on the stake weight proportion and slot probability.
+pub fn calculate_bundle_election_threshold(
+    stake_weight: StakeWeight,
+    total_stake_weight: StakeWeight,
+    slot_probability: (u64, u64),
+) -> u128 {
+    // The calculation is written for not causing the overflow, which might be harder to
+    // understand, the formula in a readable form is as followes:
+    //
+    //              slot_probability.0      stake_weight
+    // threshold =  ------------------ * --------------------- * u128::MAX
+    //              slot_probability.1    total_stake_weight
+    //
+    // TODO: better to have more audits on this calculation.
+    u128::MAX / total_stake_weight * u128::from(slot_probability.0) / u128::from(slot_probability.1)
+        * stake_weight
+}
+
+pub fn is_election_solution_within_threshold(election_solution: u128, threshold: u128) -> bool {
+    election_solution <= threshold
+}
+
+/// Make a VRF transcript.
+pub fn make_local_randomness_transcript(slot_randomness: &Blake2b256Hash) -> Transcript {
+    let mut transcript = Transcript::new(VRF_TRANSCRIPT_LABEL);
+    transcript.append_message(b"slot randomness", slot_randomness);
+    transcript
+}
+
+/// Make a VRF transcript data.
+#[cfg(feature = "std")]
+pub fn make_local_randomness_transcript_data(
+    slot_randomness: &Blake2b256Hash,
+) -> VRFTranscriptData {
+    VRFTranscriptData {
+        label: VRF_TRANSCRIPT_LABEL,
+        items: vec![(
+            "slot randomness",
+            VRFTranscriptValue::Bytes(slot_randomness.to_vec()),
+        )],
+    }
+}
+
+pub mod well_known_keys {
```

Comment:
We can do it on the secondary chain side, but it's not possible on the primary chain side because the primary chain pallet `pallet-executor` also needs to know the storage key to read the corresponding state in the storage proof but there is no way to access the secondary pallet info in a typed way on the primary chain side.

Reply:
Sometimes it is more efficient to have hardcoded keys, if possible I just add a test case to make sure we catch regression, but avoid runtime overhead. I believe I did that a few times already.